jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,517 @@
+     */
+    public DosHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("maxRequestsPerSecond") int maxRequestsPerSecond,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, new ExponentialMovingAverageRateControlFactory(maxRequestsPerSecond), null, maxTrackers);
+    }
+
+    /**
+     * @param getId Function to extract an remote ID from a request.
+     * @param rateControlFactory Factory to create a Rate per Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DosHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControlFactory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, rateControlFactory, rejectHandler, maxTrackers);
+    }
+
+    /**
+     * @param handler Then next {@link Handler} or {@code null}
+     * @param getId Function to extract an remote ID from a request.
+     * @param rateControlFactory Factory to create a Rate per Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DosHandler(
+        @Name("handler") Handler handler,
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControlFactory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        super(handler);
+        installBean(_trackers);
+        _getId = Objects.requireNonNullElse(getId, ID_FROM_REMOTE_ADDRESS);
+        installBean(_getId);
+        _rateControlFactory = Objects.requireNonNull(rateControlFactory);
+        installBean(_rateControlFactory);
+        _maxTrackers = maxTrackers <= 0 ? 10_000 : maxTrackers;
+        _rejectHandler = Objects.requireNonNullElseGet(rejectHandler, TooManyRequestsRejectHandler::new);
+        installBean(_rejectHandler);
+    }
+
+    @Override
+    public void setServer(Server server)
+    {
+        super.setServer(server);
+        if (_rejectHandler instanceof Handler handler)
+            handler.setServer(server);
+    }
+
+    @Override
+    protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() > _maxTrackers)
```

Comment:
Also do you also think this behaviour could make things worse for a DDOS attack, where we reach `_maxTrackers` and just reject all the valid requests even if they haven't exceeded any rate limit?

Reply:
I've added a sweep to remove idle tracker, so at least if we are rejecting we know that we really are at max capacity of tracker.

I'm not sure what else we can do. When presented with this situation we have a few choices:
1) accept the request anyway and use more Tracker memory, which is a DoS vector but can be configured with maxTrackers = -1
2) reject the new request, which is denying service to an unknown remote address.  The argument can be made it is better to give service to the clients that you do have rather than break trying to give service to everybody?
3) drop the least busy trackers, but then our tracking becomes inaccurate.

Reply:
Yeah not sure what else could be done. 

But I don't think it matters that typical loads will have under 10k clients. Because instead of the attacker scaling up the number of requests, the attacker could just scale up the number of unique remoteAddresses and cause a DOS.

Maybe you should just call this `RateLimitHandler` instead of `DosHandler`?

Also you should fix the other issue of being able to exceed the `_maxTrackers` limit which I mentioned in the first comment.